mtp_api/apps/security/models.py

```diff
@@ -0,0 +1,97 @@
+
+    @property
+    def credit_filters(self):
+        return (
+            reduce(
+                lambda x, y: x | y,
+                chain(
+                    (
+                        models.Q(transaction__sender_name=d.sender_name) &
+                        models.Q(transaction__sender_sort_code=d.sender_sort_code) &
+                        models.Q(transaction__sender_account_number=d.sender_account_number) &
+                        models.Q(transaction__sender_roll_number=d.sender_roll_number)
+                        for d in self.bank_transfer_details.all()
+                    ),
+                    (
+                        models.Q(payment__card_number_last_digits=d.card_number_last_digits) &
+                        models.Q(payment__card_expiry_date=d.card_expiry_date)
+                        for d in self.debit_card_details.all()
+                    )
+                ),
+                models.Q(pk=None)
+            )
+        )
+
+    class Meta:
+        permissions = (
+            ('view_senderprofile', 'Can view sender profile'),
+        )
+
+
+class BankTransferSenderDetails(TimeStampedModel):
+    sender_name = models.CharField(max_length=250, blank=True)
+    sender_sort_code = models.CharField(max_length=50, blank=True)
+    sender_account_number = models.CharField(max_length=50, blank=True)
+    sender_roll_number = models.CharField(max_length=50, blank=True)
+    sender = models.ForeignKey(
+        SenderProfile, on_delete=models.CASCADE, related_name='bank_transfer_details'
+    )
+
+
+class DebitCardSenderDetails(TimeStampedModel):
+    card_number_last_digits = models.CharField(max_length=4, blank=True, null=True)
+    card_expiry_date = models.CharField(max_length=5, blank=True, null=True)
+    sender = models.ForeignKey(
+        SenderProfile, on_delete=models.CASCADE, related_name='debit_card_details'
+    )
+
+
+class CardholderName(models.Model):
+    name = models.CharField(max_length=250)
+    debit_card_sender_details = models.ForeignKey(
+        DebitCardSenderDetails, on_delete=models.CASCADE,
+        related_name='cardholder_names', related_query_name='cardholder_name'
+    )
+
+
+class PrisonerProfile(TimeStampedModel):
+    prisoner_name = models.CharField(max_length=250)
+    prisoner_number = models.CharField(max_length=250)
+    prisoner_dob = models.DateField()
+    credit_count = models.IntegerField(default=0)
+    credit_total = models.IntegerField(default=0)
+
+    prisons = models.ManyToManyField(Prison, related_name='prisoners')
+    senders = models.ManyToManyField(SenderProfile, related_name='prisoners')
+
+    @property
+    def credit_filters(self):
+        return (
+            models.Q(prisoner_name=self.prisoner_name) &
+            models.Q(prisoner_dob=self.prisoner_dob)
+        )
+
+    class Meta:
+        permissions = (
+            ('view_prisonerprofile', 'Can view prisoner profile'),
+        )
+
+
+class SecurityDataUpdate(models.Model):
```

Comment:
since sender and prisoner profiles are timestamped, doesn't the `.modified` field on it suffice for this purpose? not especially against this method tho.